src/multicharge/model/eeqbc.f90

```diff
@@ -48,729 +63,900 @@ module multicharge_model_eeqbc
+      type(cache_container), intent(inout) :: cache
+      real(wp), intent(out) :: dxdr(:, :, :)
+      real(wp), intent(out) :: dxdL(:, :, :)
+
+      type(eeqbc_cache), pointer :: ptr
+
+      integer :: iat, izp, jat
+      real(wp) :: tmpdcn, tmpdqloc
+      real(wp), allocatable :: dtmpdr(:, :, :), dtmpdL(:, :, :)
+
+      call view(cache, ptr)
+      allocate (dtmpdr(3, mol%nat, mol%nat + 1), dtmpdL(3, 3, mol%nat + 1))
 
       dxdr(:, :, :) = 0.0_wp
       dxdL(:, :, :) = 0.0_wp
       dtmpdr(:, :, :) = 0.0_wp
       dtmpdL(:, :, :) = 0.0_wp
       !$omp parallel do default(none) schedule(runtime) &
-      !$omp shared(cn, dcndr, dcndL, qloc, dqlocdr, dqlocdL) &
-      !$omp shared(self, mol, tmp, dtmpdr, dtmpdL) private(iat, izp) 
+      !$omp shared(ptr, self, mol, dtmpdr, dtmpdL) &
+      !$omp private(iat, izp)
       do iat = 1, mol%nat
          izp = mol%id(iat)
-         tmp(iat) = -self%chi(izp) + self%kcnchi(izp)*cn(iat) &
-            & + self%kqchi(izp)*qloc(iat)
-
          ! CN and effective charge derivative
-         dtmpdr(:, :, iat) = self%kcnchi(izp) * dcndr(:, :, iat) + dtmpdr(:, :, iat)
-         dtmpdL(:, :, iat) = self%kcnchi(izp) * dcndL(:, :, iat) + dtmpdL(:, :, iat)
-         dtmpdr(:, :, iat) = self%kqchi(izp) * dqlocdr(:, :, iat) + dtmpdr(:, :, iat)
-         dtmpdL(:, :, iat) = self%kqchi(izp) * dqlocdL(:, :, iat) + dtmpdL(:, :, iat)
+         dtmpdr(:, :, iat) = self%kcnchi(izp)*ptr%dcndr(:, :, iat) + dtmpdr(:, :, iat)
+         dtmpdL(:, :, iat) = self%kcnchi(izp)*ptr%dcndL(:, :, iat) + dtmpdL(:, :, iat)
+         dtmpdr(:, :, iat) = self%kqchi(izp)*ptr%dqlocdr(:, :, iat) + dtmpdr(:, :, iat)
+         dtmpdL(:, :, iat) = self%kqchi(izp)*ptr%dqlocdL(:, :, iat) + dtmpdL(:, :, iat)
       end do
 
-      call gemm(dtmpdr(:, :, :mol%nat), cache%cmat(:mol%nat, :mol%nat), dxdr)
-      call gemm(dtmpdL(:, :, :mol%nat), cache%cmat(:mol%nat, :mol%nat), dxdL)
+      call gemm(dtmpdr(:, :, :mol%nat), ptr%cmat(:mol%nat, :mol%nat), dxdr)
+      call gemm(dtmpdL(:, :, :mol%nat), ptr%cmat(:mol%nat, :mol%nat), dxdL)
       !call gemv(cache%dcdr(:, :, :mol%nat), tmp(:mol%nat), xvec)
 
-
       !$omp parallel do default(none) schedule(runtime) &
-      !$omp reduction(+:dxdr, dxdL) shared(self, mol, cache, tmp) &
+      !$omp reduction(+:dxdr, dxdL) shared(self, mol, ptr) &
       !$omp private(iat, jat)
       do iat = 1, mol%nat
          do jat = 1, mol%nat
-            dxdr(:, iat, iat) = tmp(jat) * cache%dcdr(:, iat, jat) + dxdr(:, iat, iat)
-            dxdr(:, iat, jat) = (tmp(iat) - tmp(jat)) * cache%dcdr(:, iat, jat) &
+            dxdr(:, iat, iat) = ptr%xtmp(jat)*ptr%dcdr(:, iat, jat) + dxdr(:, iat, iat)
+            dxdr(:, iat, jat) = (ptr%xtmp(iat) - ptr%xtmp(jat))*ptr%dcdr(:, iat, jat) &
                & + dxdr(:, iat, jat)
-         end do 
+         end do
       end do
-   else
+   end subroutine get_xvec_derivs
+
+   subroutine get_coulomb_derivs(self, mol, cache, vrhs, dadr, dadL, atrace)
```

Comment:
This was mainly the case because vrhs was used as an intermediate storage and qvec is the result variable. Additionally some later steps need the full length of vrhs for the matrix-vector multiplication. So both is possible but in the function the variable should have a name that is understandable